src/pass/saveload_json.cc

```diff
@@ -154,86 +171,130 @@ struct JSONGraph {
   }
 };
 
-// Load a graph from JSON file.
-Graph LoadJSON(Graph src) {
-  CHECK_NE(src.attrs.count("json"), 0U)
-      << "Load JSON require json to be presented.";
-  const std::string &json_str =
-      nnvm::get<std::string>(*src.attrs.at("json"));
-  bool no_parse = false;
-  if (src.attrs.count("load_json_no_parse")) {
-    no_parse = nnvm::get<bool>(*src.attrs.at("load_json_no_parse"));
+void Symbol2JSONGraph(std::shared_ptr<Symbol> src, JSONGraph &jgraph) {
+  std::unordered_map<Node*, uint32_t> node2index;
+  jgraph.node_row_ptr.push_back(0);
+  DFSVisit(src->outputs, [&node2index, &jgraph](const NodePtr& n) {
+    uint32_t nid = static_cast<uint32_t>(jgraph.nodes.size());
+    node2index[n.get()] = nid;
+    if (n->is_variable()) {
+      jgraph.arg_nodes.push_back(nid);
+    }
+    JSONNode jnode;
+    jnode.node = n;
+    jnode.inputs.reserve(n->inputs.size());
+    for (const NodeEntry& e : n->inputs) {
+      jnode.inputs.emplace_back(node2index.at(e.node.get()), e.index, e.version);
+    }
+    for (const NodePtr& c : n->control_deps) {
+      jnode.control_deps.push_back(node2index.at(c.get()));
+    }
+    jgraph.node_row_ptr.push_back(jgraph.node_row_ptr.back() + n->num_outputs());
+    jgraph.nodes.emplace_back(std::move(jnode));
+  });
+  for (const NodeEntry& e : src->outputs) {
+    jgraph.heads.emplace_back(node2index.at(e.node.get()), e.index, e.version);
   }
-  std::istringstream is(json_str);
-  dmlc::JSONReader reader(&is);
-  JSONGraph jgraph;
-  // load in json graph.
-  jgraph.Load(&reader);
-  // connects the nodes
+}
+
+std::shared_ptr<Symbol> JSONGraph2Symbol(JSONGraph &jgraph, bool no_parse) {
```

Comment:
why do you need to change jgraph here?

Reply:
You are right. The jgraph here can be constant reference, "JSONNode &n" can be constant reference as well. Should I update this pull request?